### DISCUSSION TITLE
What is Mosaic

### DISCUSSION THREAD

**Discussion by krzysztofrusek** [jax-ml/jax#16855] [General] [answered]
I found it appearing in few recent commits and I am curious what is it?

**Comment by apaszke** [comment, marked as answer]
It's a new sub-project we've been working on and are in the progress of making all its features public. It's not fully set up to be used publicly yet, but stay tuned for the upcoming announcements 😊

**Reply by krzysztofrusek** [reply]
Thanks!

**Reply by garymm** [reply]
So a year and a half later, any updates?

**Reply by jakevdp** [reply]
See https://docs.jax.dev/en/latest/pallas/index.html

**Reply by garymm** [reply]
From skimming the docs, I guess Mosaic is a compiler and / or IR (not sure which) that targets TPUs and Nvidia GPUs. Unclear how it relates to XLA.

**Reply by apaszke** [reply]
Yes, that description is fair. Mosaic is a compiler (and an input IR it can compile) that we use for the Pallas TPU backend and the experimental new Pallas GPU backend. While conceptually it is possible to use Mosaic without Pallas, for as long as you're interested in using JAX, we'd highly recommend using Pallas instead. It has a much more stable API and overall is more convenient to use, and it does not really come with any downsides. Anything expressible in Mosaic directly should be expressible in Pallas too.

**Comment by Lekja00160612** [comment]
I have the same question, does it have anything to do with MosaicML or just jaxlib related internal?

**Reply by hawkinsp** [reply]
It has nothing to do with MosaicML.
